tool/tctl/sso/tester/command.go

```diff
@@ -0,0 +1,266 @@
+
+	cfg := client.MakeDefaultConfig()
+	cfg.WebProxyAddr = proxies[0].GetPublicAddr()
+
+	tc, err := client.NewClient(cfg)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return client.SSHAgentSSOLogin(ctx, client.SSHLoginSSO{
+		SSHLogin: client.SSHLogin{
+			ProxyAddr:         tc.WebProxyAddr,
+			PubKey:            key.Pub,
+			TTL:               tc.KeyTTL,
+			Insecure:          tc.InsecureSkipVerify,
+			Pool:              nil,
+			Compatibility:     tc.CertificateFormat,
+			RouteToCluster:    tc.SiteName,
+			KubernetesCluster: tc.KubernetesCluster,
+		},
+		ConnectorID: "-sso-test",
+		Protocol:    protocol,
+		BindAddr:    tc.BindAddr,
+		Browser:     tc.Browser,
+	}, config)
+}
+
+type DiagnosticInfoField struct {
+	Present bool
+	Show    bool
+	Msg     string
+}
+
+func (cmd *SSOTestCommand) reportLoginResult(authKind string, diag *types.SSODiagnosticInfo, infoErr error, loginResponse *auth.SSHLoginResponse, loginErr error) (errResult error) {
+	success := diag != nil && diag.Success
+
+	// check for errors
+	if loginErr != nil || infoErr != nil {
+		success = false
+	}
+
+	if success {
+		fmt.Printf("Success! Logged in as: %v\n", loginResponse.Username)
+	} else {
+		fmt.Printf("Failure!\n")
+		errResult = trace.Errorf("SSO flow failed.")
+
+		if infoErr != nil {
+			fmt.Printf("No diagnostic info found. Most likely cause: the request timed out or callback configuration is incorrect. Ensure that user logs within alloted time and IdP configuration is correct.\n Error details: %v\n", trace.UserMessage(infoErr))
+			errResult = trace.Wrap(infoErr, "SSO flow failed.")
+		}
+
+		if loginErr != nil {
+			fmt.Printf("Login error: %v\n", trace.UserMessage(loginErr))
+			errResult = trace.Wrap(loginErr, "SSO flow failed.")
+		}
+	}
+
+	// finish early if there is no diag info to show.
+	if diag == nil {
+		return errResult
+	}
+
+	fields := []DiagnosticInfoField{
+		// common fields across auth connector types.
+		{
+			Present: diag.Error != "",
+			Show:    cmd.config.Debug || loginErr == nil,
+			Msg:     FormatString("Original error", diag.Error),
+		},
+		{
+			Present: diag.CreateUserParams != nil,
+			Show:    true,
+			Msg:     formatUserDetails("Authentication details", diag.CreateUserParams),
+		},
+	}
+
+	// enrich the fields with auth-specific fields
+	if getFields, ok := cmd.GetDiagInfoFields[authKind]; ok {
+		fields = append(fields, getFields(diag, cmd.config.Debug)...)
+	}
+
+	// we want this field last.
+	// raw data - debug
+	fields = append(fields,
+		DiagnosticInfoField{
+			Present: true,
+			Show:    cmd.config.Debug,
+			Msg:     FormatJSON("Raw data", diag),
+		})
+
+	const termWidth = 80
+
+	for _, f := range fields {
+		if f.Present && f.Show {
+			fmt.Println(strings.Repeat("-", termWidth))
+			fmt.Println(f.Msg)
+		}
+	}
+
+	if !cmd.config.Debug {
+		fmt.Println(strings.Repeat("-", termWidth))
+		fmt.Println("For more details repeat the command with --debug flag.")
+	}
+
+	fmt.Println()
+
+	return errResult
```

Comment:
The large block of code between the creation of errResult make the code hard to read 
Can we extract the logic to a separate function and reduce call: 
```suggestion
	if diag != nil {
		printDiagnosticInfo(diag, ...)
	}
	return errResult

```

Reply:
Done, reads better indeed.